lib/fs/basicfs.go

```diff
@@ -352,3 +353,11 @@ func (e *ErrWatchEventOutsideRoot) Error() string {
 func (f *BasicFilesystem) newErrWatchEventOutsideRoot(absPath string, roots []string) *ErrWatchEventOutsideRoot {
 	return &ErrWatchEventOutsideRoot{fmt.Sprintf("Watching for changes encountered an event outside of the filesystem root: f.root==%v, roots==%v, path==%v. This should never happen, please report this message to forum.syncthing.net.", f.root, roots, absPath)}
 }
+
+func (f *BasicFilesystem) WriteFile(name string, content []byte, perm FileMode) error {
+	name, err := f.rooted(name)
```

Comment:
Hell no. This doesn't make sense to have on the filesystem interface.

Reply:
It's not on the `Filesystem` interface. I don't know why `fs.BasicFilesystem` is exported, but it's only used as `fs.Filesystem`. This is exposed as a utility function, i.e. `fs.WriteFile(fs.Filesystem, ...)`, which is a replacement for `ioutil.WriteFile`. It's in the filesystem package because of cache invalidation and as it seems like the logical place for it to go.

Reply:
We use this in tests only no? If yes, it should stay in tests, and we should not expand public interface for tests.

Reply:
We do use ioutil.WriteFile in api, which isn't "filesystemified". It just felt like a natural extension that might be useful earlier or later anyway so why not have it. However I agree as it's not needed now, it's better to have just a testing util function instead.

Reply:
WriteFile is not a native filesystem construct so it should not be part of the contract. You don't have os.WriteFile, you have ioutil.WriteFile, which is some helper package.